### ISSUE TITLE
Queries on views are slow

### ISSUE THREAD

**Issue by jmcarp** [18F/openFEC#540]
Because the logic behind the views can get pretty complicated, even simple queries on the views are noticeably slow--getting a count on the `Candidates` view took about 375 ms on my laptop, and there are only about 1000 candidates. If data are going to be imported periodically and won't be streaming in constantly, or if it's tolerable to slow slightly stale data, then this looks like a good use case for materialized views, which are cached on disk and don't have to be rebuilt for every query.

I'll try this out and compare performance unless anybody has reservations @catherinedevlin @LindsayYoung @arowla 

FWIW, here's the query plan for a simple count on `Candidates`:

```
                                                                                       QUERY PLAN                                                                                       
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=1267.82..1267.83 rows=1 width=0) (actual time=366.302..366.302 rows=1 loops=1)
   ->  GroupAggregate  (cost=252.37..1267.61 rows=17 width=110) (actual time=17.655..366.015 rows=1044 loops=1)
         Group Key: dimcand.cand_sk, dimcand.cand_id, csi_recent.election_yr
         ->  Sort  (cost=252.37..252.41 rows=17 width=110) (actual time=16.779..17.040 rows=1162 loops=1)
               Sort Key: dimcand.cand_sk, dimcand.cand_id, csi_recent.election_yr
               Sort Method: quicksort  Memory: 212kB
               ->  Hash Join  (cost=227.85..252.02 rows=17 width=110) (actual time=8.799..14.988 rows=1162 loops=1)
                     Hash Cond: (co.party_sk = dimparty.party_sk)
                     ->  Nested Loop  (cost=226.36..250.29 rows=17 width=54) (actual time=8.724..14.113 rows=1162 loops=1)
                           ->  Hash Right Join  (cost=226.08..244.83 rows=17 width=48) (actual time=8.709..9.851 rows=1162 loops=1)
                                 Hash Cond: (csi_all.cand_sk = dimcand.cand_sk)
                                 ->  Seq Scan on dimcandstatusici csi_all  (cost=0.00..15.49 rows=849 width=13) (actual time=0.005..0.145 rows=849 loops=1)
                                 ->  Hash  (cost=225.87..225.87 rows=17 width=43) (actual time=8.685..8.685 rows=1157 loops=1)
                                       Buckets: 1024  Batches: 1  Memory Usage: 85kB
                                       ->  Hash Left Join  (cost=146.56..225.87 rows=17 width=43) (actual time=4.380..7.909 rows=1157 loops=1)
                                             Hash Cond: (dimcand.cand_sk = csi_recent.cand_sk)
                                             Filter: ((csi_recent.election_yr IS NULL) OR ((co.cand_election_yr)::numeric = csi_recent.election_yr))
                                             Rows Removed by Filter: 544
                                             ->  Hash Join  (cost=66.53..129.06 rows=1701 width=38) (actual time=2.155..4.128 rows=1701 loops=1)
                                                   Hash Cond: (co.cand_sk = dimcand.cand_sk)
                                                   ->  Seq Scan on dimcandoffice co  (cost=0.00..37.01 rows=1701 width=28) (actual time=0.011..0.463 rows=1701 loops=1)
                                                   ->  Hash  (cost=40.68..40.68 rows=2068 width=18) (actual time=2.122..2.122 rows=2068 loops=1)
                                                         Buckets: 1024  Batches: 1  Memory Usage: 101kB
                                                         ->  Seq Scan on dimcand  (cost=0.00..40.68 rows=2068 width=18) (actual time=0.017..0.947 rows=2068 loops=1)
                                             ->  Hash  (cost=69.48..69.48 rows=844 width=17) (actual time=2.136..2.136 rows=844 loops=1)
                                                   Buckets: 1024  Batches: 1  Memory Usage: 44kB
                                                   ->  Subquery Scan on csi_recent  (cost=56.79..69.48 rows=844 width=17) (actual time=1.216..1.787 rows=844 loops=1)
                                                         ->  Unique  (cost=56.79..61.04 rows=844 width=17) (actual time=1.215..1.557 rows=844 loops=1)
                                                               ->  Sort  (cost=56.79..58.91 rows=849 width=17) (actual time=1.214..1.299 rows=849 loops=1)
                                                                     Sort Key: dimcandstatusici.cand_sk, dimcandstatusici.election_yr
                                                                     Sort Method: quicksort  Memory: 79kB
                                                                     ->  Seq Scan on dimcandstatusici  (cost=0.00..15.49 rows=849 width=17) (actual time=0.026..0.419 rows=849 loops=1)
                           ->  Index Scan using dimoffice_pkey on dimoffice  (cost=0.27..0.31 rows=1 width=22) (actual time=0.003..0.003 rows=1 loops=1162)
                                 Index Cond: (office_sk = co.office_sk)
                     ->  Hash  (cost=1.22..1.22 rows=22 width=72) (actual time=0.038..0.038 rows=22 loops=1)
                           Buckets: 1024  Batches: 1  Memory Usage: 2kB
                           ->  Seq Scan on dimparty  (cost=0.00..1.22 rows=22 width=72) (actual time=0.011..0.017 rows=22 loops=1)
         SubPlan 1
           ->  Limit  (cost=59.62..59.63 rows=1 width=26) (actual time=0.323..0.323 rows=1 loops=1044)
                 ->  Sort  (cost=59.62..59.63 rows=1 width=26) (actual time=0.322..0.322 rows=1 loops=1044)
                       Sort Key: cp.candproperties_sk
                       Sort Method: quicksort  Memory: 25kB
                       ->  Seq Scan on dimcandproperties cp  (cost=0.00..59.61 rows=1 width=26) (actual time=0.159..0.318 rows=1 loops=1044)
                             Filter: (cand_sk = dimcand.cand_sk)
                             Rows Removed by Filter: 1648
 Planning time: 5.586 ms
 Execution time: 366.764 ms
(47 rows)
```

**Comment by arowla** [comment]
Most of our views are already materialized. If you see one that isn't and ought to be, be my guest. :) Just please rename the table suffix to "_mv" so it can be quickly differentiated (all our MV's may not fit this pattern, but trying to get more organized about that...)

**Comment by arowla** [comment]
Oh, and our data will be updated nightly at best. So yes, this is the perfect use case for them.

**Comment by LindsayYoung** [comment]
are 35,739 candidates the endpoint defaults to the last few years, but you can see all candidates by looking at  `/candidates?year=*`

**Comment by jmcarp** [comment]
Got it, thanks for help! I didn't realize most of the views were already materialized. Will send a patch shortly.

@LindsayYoung I'm probably looking in the wrong place, but I got about 1000 candidates locally like this:
```
select count(*) from ofec_candidates_mv;
```
Should there be more?

**Comment by LindsayYoung** [comment]
No worries- I thought you were referring to the API's candidates view and not ofec_candidates in the test dataset.

**Comment by jmcarp** [comment]
Addressed in #545.
